trin-core/src/types/block_body.rs

```diff
@@ -0,0 +1,307 @@
+        if transaction.is_empty() {
+            // at least one byte needs to be present
+            return Err(DecoderError::RlpIncorrectListLen);
+        }
+        let id = TransactionId::try_from(transaction[0]);
+        if id.is_err() {
+            return Err(DecoderError::Custom("Unknown transaction type"));
+        }
+        match id.unwrap() {
+            TransactionId::EIP1559 => Ok(Self::EIP1559(rlp::decode(&transaction[1..])?)),
+            TransactionId::AccessList => Ok(Self::AccessList(rlp::decode(&transaction[1..])?)),
+            TransactionId::Legacy => Ok(Self::Legacy(rlp::decode(transaction)?)),
+        }
+    }
+
+    pub fn encode(&self) -> Vec<u8> {
+        let mut stream = RlpStream::new();
+        match self {
+            Self::Legacy(tx) => {
+                tx.rlp_append(&mut stream);
+                stream.out().freeze().to_vec()
+            }
+            Self::AccessList(tx) => {
+                tx.rlp_append(&mut stream);
+                [&[TransactionId::AccessList as u8], stream.as_raw()].concat()
+            }
+            Self::EIP1559(tx) => {
+                tx.rlp_append(&mut stream);
+                [&[TransactionId::EIP1559 as u8], stream.as_raw()].concat()
+            }
+        }
+    }
+}
+
+#[derive(Default, Debug, Clone, PartialEq, Eq, RlpEncodable, RlpDecodable)]
+pub struct LegacyTransaction {
+    pub nonce: U256,
+    pub gas_price: U256,
+    pub gas: U256,
+    pub to: Address,
+    pub value: U256,
+    pub data: Bytes,
+    pub v: U64,
+    pub r: U256,
+    pub s: U256,
+}
+
+#[derive(Eq, Debug, Clone, PartialEq, RlpDecodable, RlpEncodable)]
+pub struct AccessListTransaction {
+    pub chain_id: U256,
+    pub nonce: U256,
+    pub gas_price: U256,
+    pub gas_limit: U256,
+    pub to: H160,
+    pub value: U256,
+    pub data: Bytes,
+    pub access_list: AccessList,
+    pub y_parity: U64,
+    pub r: U256,
+    pub s: U256,
+}
+
+#[derive(Eq, Debug, Clone, PartialEq, RlpDecodable, RlpEncodable)]
+pub struct EIP1559Transaction {
+    pub chain_id: U256,
+    pub nonce: U256,
+    pub max_priority_fee_per_gas: U256,
+    pub max_fee_per_gas: U256,
+    pub gas_limit: U256,
+    pub to: H160,
+    pub value: U256,
+    pub data: Bytes,
+    pub access_list: AccessList,
+    pub y_parity: U64,
+    pub r: U256,
+    pub s: U256,
+}
+
+#[derive(Default, Debug, PartialEq, Eq, Clone)]
+pub struct AccessList {
+    pub list: Vec<AccessListItem>,
+}
+
+impl Decodable for AccessList {
+    fn decode(rlp: &Rlp) -> Result<Self, DecoderError> {
+        let list = rlp
+            .iter()
+            .map(|v| rlp::decode(v.as_raw()).unwrap())
+            .collect();
+        Ok(Self { list })
+    }
+}
+
+impl Encodable for AccessList {
+    fn rlp_append(&self, stream: &mut RlpStream) {
+        stream.begin_list(self.list.len());
+        for access in self.list.iter() {
+            stream.append(access);
+        }
```

Comment:
This is the same, right?
```suggestion
        stream.append_list(&self.list);
```
or does that not work with a list of encodable items?